### PR TITLE
Add asserts to cuSTL containers

### PR DESCRIPTION
**Pull request by Heikman**
Add asserts to cuSTL containers checking sizes at assignment.

@Flamefire 

solves #1194

### CODE REVIEW

**Comment by Flamefire** [comment]
`assert()` is only evaluated in debug builds. I'd at least make this a runtime error (e.g. throw std::invalid_argument) and mention the size restriction in the documentation.
Besides: You use the boost macro in 2 places and plain `assert` in 2 other in a single PR. Pretty inconsistent.

**Comment by Heikman** [comment]
Now, the assert is suppressed on device because first it is very expensive yet and second the assign operator is even just called hypothetically on device so far.

**Comment by Flamefire** [comment]
I'd still use runtime errors instead of asserts.
You can still use an `assert` (plain version) for the device side of `HDINLINE` functions (only `CartBuffer`)

**Comment by Heikman** [comment]
done.

**Comment by Flamefire** [comment]
Small issue: Boost headers should be included before STL headers
Big issue: Still no runtime checks. `BOOST_ASSERT` is equal to `assert` but can be ***disabled*** separately.

**Comment by Heikman** [comment]
done.

**Comment by psychocoderHPC** [comment]
@Flamefire: Did you accept the last changes that `assert` is only used on device?

File: src/libPMacc/include/cuSTL/container/DeviceBuffer.hpp [hunk 1]
```diff
@@ -23,6 +23,7 @@
 
 #pragma once
 
+#include <boost/assert.hpp>
```
Comment (Flamefire) [inline src/libPMacc/include/cuSTL/container/DeviceBuffer.hpp]: Please move this down before the std includes

File: src/libPMacc/include/cuSTL/container/HostBuffer.hpp [hunk 2]
```diff
@@ -22,6 +22,7 @@
 
 #pragma once
 
+#include <boost/assert.hpp>
```
Comment (Flamefire) [inline src/libPMacc/include/cuSTL/container/HostBuffer.hpp]: Move down

**Comment by Flamefire** [comment]
Yes it is ok since there are no exceptions on the device AFAIK and this is better overhead-wise.
However I'd add a temporary for the stringstream which makes the code cleaner and removes the typecast.
Also 2 small issues with include order

**Comment by psychocoderHPC** [comment]
@Heikman  please move the include marked by @Flamefire down, after that I can merge this pull request

**Comment by Heikman** [comment]
done

**Pull request merged**